api/v1alpha1/wasm_types.go

```diff
@@ -0,0 +1,66 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Wasm defines a wasm extension.
+type Wasm struct {
+	// Name is a unique name for this Wasm extension. It is used to identify the
+	// Wasm extension if multiple extensions are handled by the same vm_id and root_id.
+	// It's also used for logging/debugging.
+	Name string `json:"name"`
+
+	// VmID is an ID that will be used along with a hash of the wasm code to
+	// determine which VM will be used to load the Wasm extension. All extensions
+	// that have the same vm_id and code will use the same VM.
+	//
+	// Note that sharing a VM between plugins can reduce memory utilization and
+	// make sharing of data easier, but it may have security implications.
+	VmID *string `json:"vmID,omitempty"`
+
+	// RootID is a unique ID for a set of extensions in a VM which will share a
+	// RootContext and Contexts if applicable (e.g., an Wasm HttpFilter and an Wasm AccessLog).
+	// If left blank, all extensions with a blank root_id with the same vm_id will share Context(s).
+	RootID *string `json:"rootID,omitempty"`
+
+	// Code is the wasm code for the extension.
+	Code WasmCodeSource `json:"code"`
+
+	// Config is the configuration for the Wasm extension.
+	// This configuration will be passed as a JSON string to the Wasm extension.
+	Config *apiextensionsv1.JSON `json:"config"`
```

Comment:
Envoy supports struct, string, and bytes. Follow Istio practice to restrict input as a JSON string for a better UI:

* We can avoid media disruption in the K8s resources by just specifying the config as a YAML object. JSON is a subset of YAML, so it is very easy to convert to it from YAML.
* JSON has parsers in almost every language (as opposed to e.g. YAML), and certainly the ones that already compile to Wasm. It is also compatible with protobuf, should we want to use that in some of our SDKs.
* Developers don’t have to deal with parsing configuration. The language SDKs can use optimized code to unmarshall the JSON configuration and make the values easily accessible to developers in a language-specific way (this could e.g. be a map[string]interface{} in golang).
* Using a structured format for configuration allows for configuration injection and templating


https://docs.google.com/document/d/15GQnhOpLfvgyOZ8VDF5KD-q2cjXEMLdTQMm0YPswIu0/edit#heading=h.6zqg8n4aice5